unbound refactor the unit & add systemd support

andir opened the pull request:
##### Motivation for this change

This is motivated due to the current nasty state of the module and the uncertain startup state unbound might be in during boot-up on NixOS. It is somewhat related to #99901 as that might improve the situation around ACME during systemd switches (where the unbound unit was changed).

###### Things to be done

- [x] write a NixOS VM test to verify it still works
- [x] add a symlink pointing to the conf at `/etc/unbound/unbound.conf`. This will improve the `unbound-control` usability, we won't need to explicitly point to the config through the `-c` arg.
- [x] provide a module system option for the control socket (it should default to a root/unbound only unix socket; if configured)
- [x] write changelog entries
- [ ] remove the remaining FIXME's and squash some more.

andir committed:
unbound: allow building with systemd support

Systemd has to remain an optional (non-default) dependency as otherwise
we will have an unpleasant bootstrap cycle. Most (if not all) of the
(lib)unbound consumers will likely not care about unbound's systemd
integration that only affects the daemon mode, anyway.

andir committed:
unbound-with-systemd: init

This introduces an unbound variant that is built with systemd support.
That means it is able to signal readiness to systemd once it did start
or finished reloading. This likely allows us to close a small gap during
bootup where the service is assumed up but doesn't respond to DNS
queries just yet.

andir committed:
nixos/modules/services/networking/unbound: update systemd unit

Previously we just applied a very minimal set of restrictions and
trusted unbound to properly drop root privs and capabilities.

With this change I am (for the most part) just using the upstream
example unit file for unbound. The main difference is that we start
unbound was `unbound` user with the required capabilities instead of
letting unbound do the chroot & uid/gid changes.

The upstream unit configuration this is based on is a lot stricter with
all kinds of permissions then our previous variant. It also came with
the default of having the `Type` set to `notify`, therefore we are also
using the `unbound-with-systemd` package here. Unbound will start up,
read the configuration files and start listening on the configured ports
before systemd will declare the unit "running". This will likely help
with startup order and the occasional race condition during system
activation where the DNS service is started but not yet ready to answer
queries.

Aditionally to the much stricter runtime environmet I removed the
`/dev/urandom` mount lines we previously had in the code (that would
randomly fail during `stop`-phase).

The `preStart` script is now only required if we enabled the trust
anchor updates (which are still enabled by default).

Another beneefit of the refactoring is that we can now issue reloads via
either `pkill -HUP unbound` or `systemctl reload unbound` to reload the
running configuration without taking the daemon offline. A prerequisite
of this was that unbound configuration is available on a well known path
on the file system. I went for /etc/unbound/unbound.conf as that is the
default in the CLI tooling which in turn enables us to use
`unbound-control` without passing a custom configuration location.

andir committed:
nixos/tests/unbound: init

flokli commented:
Generally I agree with this approach - I'd personally prefer calling the library variant `libunbound`, and `unbound-systemd` `unbound` (like we do with `[lib]pulseaudio`).

However, the test currently fails:
```
resolver: must succeed: drill  -t AAAA example.com @localhost
resolver # Error: error sending query: General LDNS error
resolver: output:
error:
Traceback (most recent call last):
  File "/nix/store/isl6j0894s1kyp4l9y6wjpnmfdc1f7n5-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 901, in run_tests
    exec(tests, globals())
  File "<string>", line 1, in <module>
  File "<string>", line 31, in <module>
  File "<string>", line 17, in test
  File "/nix/store/isl6j0894s1kyp4l9y6wjpnmfdc1f7n5-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 423, in succeed
    raise Exception(
Exception: command `drill  -t AAAA example.com @localhost` failed (exit code 1)
cleaning up
```

andir commented:
> Generally I agree with this approach - I'd personally prefer calling the library variant `libunbound`, and `unbound-systemd` `unbound` (like we do with `[lib]pulseaudio`).
> 
> However, the test currently fails:
> 
> ```
> resolver: must succeed: drill  -t AAAA example.com @localhost
> resolver # Error: error sending query: General LDNS error
> resolver: output:
> error:
> Traceback (most recent call last):
>   File "/nix/store/isl6j0894s1kyp4l9y6wjpnmfdc1f7n5-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 901, in run_tests
>     exec(tests, globals())
>   File "<string>", line 1, in <module>
>   File "<string>", line 31, in <module>
>   File "<string>", line 17, in test
>   File "/nix/store/isl6j0894s1kyp4l9y6wjpnmfdc1f7n5-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 423, in succeed
>     raise Exception(
> Exception: command `drill  -t AAAA example.com @localhost` failed (exit code 1)
> cleaning up
> ```

Yes, it is a WIP and I just headed to bed after pushing the latest revision.

andir committed:
nixos/tests/unbound: add test to verify control sockets work

andir commented:
Let's see if tests also pass on CI and not just locally.
@ofborg test unbound

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/make-acme-renew-systemd-service-depend-on-dns-nss-lookup/7412/8

andir commented:
Rebased & started cleaning up the commit messages so the release note changes can mostly be copy & pasted from those.

andir committed:
nixos/unbound: note about the AmbientCapabilities

andir committed:
nixos/unbound: add option to configure the local control socket path

This option allows users to specify a local UNIX control socket to
"remote control" the daemon. System users, that should be permitted to
access the daemon, must be in the `unbound` group in order to access the
socket. When a socket path is configured we are also creating the
required group.

Currently this only supports the UNIX socket mode while unbound actually
supports more advanced types. Users are still able to configure more
complex scenarios via the `extraConfig` attribute.

When this option is set to `null` (the default) it doesn't affect the
system configuration at all. The unbound defaults for control sockets
apply and no additional groups are created.

andir committed:
nixos/unbound: drop ReadWritePaths from systemd unit configuration

Both of the configured paths should be implicit due to RuntimeDirectory
& StateDirectory.

andir commented:
There is now an option to set the socket path. cc @NinjaTrappeur could you have a look if this is what you'd want?

NinjaTrappeur commented on nixos/modules/services/networking/unbound.nix:
```diff
@@ -86,6 +91,28 @@ in
 
+      localControlSocketPath = mkOption {
+        default = null;
+        # FIXME: What is the proper type here so users can specify strings,
```
I'm fine with keeping the current `types.str` alone. `types.path` can be relative to the $pwd at evaluation time. While it's what you want to point to a file living in your source tree, I always find awkward using this type to point to a runtime path.

andir replied:
Same but often my taste doesn't seem to follow what some in the community are pushing for.

picnoir reviewed (commented):
Yes. There's even a test, that's perfect!

andir committed:
nixos/unbound: add release notes for the changes that were introduced

As part of this patch series a few changes have been made to the unbound
serivce the deserve proper documentation.

andir commented:
I decided to cut the scope of this PR and leave the openresolv configuration for now. It ended up to be a much bigger rabbit hole then I initially expected. I would also like to postpone the renaming of `unbound` to `libunbound` to another PR as those are more likely to introduce breakage. I can do one right after this is merged.

Compared the the previous version I now have written the release notes entries. The rendered version can be found at: https://s.rammhold.de/nixos-manual-unbound/release-notes.html#sec-release-21.03-notable-changes

I am including a screenshot of it for the lazy people :)

![rn](https://user-images.githubusercontent.com/638836/97987304-553c7180-1ddb-11eb-92c3-afae7b29b94a.png)

The new `localControlSocketPath` option documentation can be found here: https://s.rammhold.de/nixos-manual-unbound/options.html#opt-services.unbound.localControlSocketPath

![opt](https://user-images.githubusercontent.com/638836/97987415-8a48c400-1ddb-11eb-8200-7d2c0913b4ae.png)


I am marking this as ready for review and would appreciate any feedback.

andir committed:
nixos/unbond: unbound should be required for nss-lookup.target

Other units depend on nss-lookup.target and expect the DNS resolution to
work once that target is reached. The previous version
`wants=nss-lookup.target` made this unit require the nss-lookup.target
to be reached before this was started.

Another change that we can probalby do is drop the before relationship
with the nss-lookup.target. That might just be implied with the current
version.

picnoir reviewed (approved):
:+1: The API barely moves, the module quality rise quite significantly.

Great job!

Overall LGTM.

I was wondering how to plug a `passthru.tests` to unbound. It'd be something nice to have. Sadly, I did not find any solution.

Obviously, we'd need the systemd support to run the VM test, which default to false in `pkgs/tools/networking/unbonud/default.nix`. I don't see any way to automatically run the VM test on a `unbound/default.nix` version update :( At least not a way to force ofborg to re-evaluate the test. Maybe I'm missing something.

rissson commented:
Ugh this conflicts a lot with #89572 with I caught it sooner.

NinjaTrappeur commented:
Ugh, missed it as well.

On the other hand, the test is going to make your PR more easy to test.

hmenke commented:
Backport to 20.09?

andir commented:
On 12:02 09.11.20, Henri Menke wrote:
> Backport to 20.09?

While I would like that I think we should not. The API did not really
change but the implementation did change. Anyone upgrading their 20.09
system would not expect such a "breaking" change that required lengthy
description in the changelog (for 21.03).

rissson commented on nixos/tests/unbound.nix:
```diff
@@ -0,0 +1,278 @@
+        };
+
+        environment.etc = {
+          "unbound-extra1.conf".text = ''
```
Shouldn't this have been `unbound/extra1.conf` ?

andir replied:
No, they are being symlinked to that path (`/etc/unbound/extra1.conf`) during runtime to test runtime reconfiguration.

andir replied:
See: https://github.com/NixOS/nixpkgs/pull/101218/files/5903ea539513c8ec2a1412d88465559b957b6483#diff-86138e3965f58677a4f62dadc7dd5a42fc8054f1a5edab23f3745d8aa21506fbR250

kvtb commented on nixos/modules/services/networking/unbound.nix:
```diff
@@ -31,19 +29,25 @@ let
       ${interfaces}
       ${access}
       ${trustAnchor}
+    ${lib.optionalString (cfg.localControlSocketPath != null) ''
+      remote-control:
+        control-enable: yes
+        control-interface: ${cfg.localControlSocketPath}
+    ''}
     ${cfg.extraConfig}
```
Introducing `cfg.localControlSocketPath` made impossible to write `do`-options (like `do-ip6: no`) in `cfg.extraConfig`, because those options must be BEFORE `remote-control`

AluisioASG replied:
You can just begin a new `server:` block in `extraConfig`, at least I used to run a config like that.

kvtb commented on nixos/modules/services/networking/unbound.nix:
```diff
@@ -106,43 +135,85 @@ in
+        MemoryDenyWriteExecute = true;
+        NoNewPrivileges = true;
         PrivateDevices = true;
-        Restart = "always";
-        RestartSec = "5s";
```
Why removed?
`unbound` could crash (it happens) or just being started too slow that `systemd` marks the service as failed
